django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+            filter_q.add(Q(package__owner__name=params["team_identifier"]), Q.AND)
+
+        if params.get("user_identifier", None):
+            filter_q.add(
+                Q(
+                    package__owner__members__user__username__icontains=params[
+                        "user_identifier"
+                    ]
+                ),
+                Q.AND,
+            )
+
+        if params.get("package_identifier", None):
+            filter_q.add(Q(package__name=params["package_identifier"]), Q.AND)
+
+        if not params.get("deprecated", None):
+            exclude_q.add(Q(package__is_deprecated=True), Q.OR)
+
+        if not params.get("nsfw", None):
+            exclude_q.add(Q(has_nsfw_content=True), Q.OR)
+
+        if params.get("q", None):
+            icontains_query = Q()
+            parts = [x for x in params["q"].split(" ") if x]
+
+            for part in parts:
+                for field in (
+                    "package__name",
+                    "package__owner__name",
+                    "package__latest__description",
+                ):
+                    icontains_query &= ~Q(**{f"{field}__icontains": part})
+            exclude_q.add(icontains_query, Q.OR)
+
+        # Add category filters/excludes
+        included = []
+        excluded = []
+        if params.get("section", None):
+            try:
+                section = PackageListingSection.objects.prefetch_related(
+                    "require_categories", "exclude_categories"
+                ).get(slug=params["section"])
+            except PackageListingSection.DoesNotExist:
+                pass
+            else:
+                [included.append(s.slug) for s in section.require_categories.all()]
+                [excluded.append(s.slug) for s in section.exclude_categories.all()]
+
+        if params.get("included_categories", None):
+            included = list(chain(included, params["included_categories"]))
+
+        if params.get("excluded_categories", None):
+            excluded = list(chain(excluded, params["excluded_categories"]))
+
+        if len(included) > 0:
+            included_cats_q = Q()
+            for cat in sorted(set(included)):
+                included_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
+            filter_q.add(included_cats_q, Q.AND)
+
+        if len(excluded) > 0:
+            excluded_cats_q = Q()
+            for cat in sorted(set(excluded)):
+                excluded_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
+            exclude_q.add(excluded_cats_q, Q.OR)
+
+        return (
+            PackageListing.objects.active()
+            .filter(
+                review_status__icontained_by=Case(
+                    When(
+                        Q(community__require_package_listing_approval=True),
+                        then=Value("approved"),
+                    ),
+                    default=Value("unreviewed approved"),
+                )
+            )
```

Comment:
Why is this filtered so oddly compared to how it's done in the rest of the project; where did the requirement come from?

Reply:
It's done in this fashion, so that it's only a single query. Not sure if it was an requirement. I probably went a bit overboard with this one, but it was fun to figure out 🐑

Reply:
I don't know whether it makes sense or not without testing, but I'd guess this to be premature optimization at this point. The performance impact would have to be pretty significant for the extra complexity to be worth it IMO.

Reply:
I have a feeling the performance hit is quite big. Whilst implementing tests in local dev, it performed a lot quicker with this, compared to the old queries 🤷